Update precompute logic to skip autograd backend when there's direction

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44361 Update fallback kernel for Autograd keys.
* #44556 Support Math keyword in native_functions.yaml.
* #44354 Add alias dispatch key Math.
* **#44349 Update precompute logic to skip autograd backend when there's direction**

registration to backend

Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

Commit:
Update fallback kernel for Autograd keys.

[ghstack-poisoned]

ailzhang commented:
Discussion: @ezyang @bhosmer with this PR I feel it's good to enforce registering a fallback for all autograd keys in core pytorch so that they get proper error messages. 
If users forget (and we don't enforce it) to register a proper fallback to AutogradPrivateUse1, they might skip the throw and get "unexpected" behavior from dispatcher. 
Does it hurt if we allow overriding fallback kernels from users so that we can set this in-tree and users can optionally override it?

Commit:
Update on "Update fallback kernel for Autograd keys."


Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

[ghstack-poisoned]

Commit:
Update on "Update fallback kernel for Autograd keys."


Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

[ghstack-poisoned]

Commit:
Update on "Update precompute logic to skip autograd backend when there's direction"


registration to backend

Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

[ghstack-poisoned]

ezyang commented:
Test?

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -211,11 +213,20 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
+    updateDispatchTableEntry_(dispatcher, k);
+  }
+  // Registering to backend key might affect computed entry at its Autograd backend key due to 2.2.
+  DispatchKey autograd_key = getAutogradKeyFromBackend(dispatch_key);
```
What if `autograd_key` is undefined?

ailzhang replied:
autograd_key is default to AutogradOther if it misses all the switch conditions ;)

ezyang commented:
The logic looks right but I'd feel better if we were exercising it somewhere, in some way.

Commit:
Update on "Update precompute logic to skip autograd backend when there's direction"


registration to backend

Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

[ghstack-poisoned]

ezyang commented on test/test_dispatch.py:
```diff
@@ -337,6 +337,73 @@ def extract_entries(table, dispatch_keys):
+AutogradOther: default_def_name_t_t [autograd catch all]
+AutogradCPU: fallthrough registered at ../aten/src/ATen/core/VariableFallbackKernel.cpp:39 [backend fallback]
+AutogradCUDA: default_def_name_t_t [autograd catch all]
+AutogradXLA: default_def_name_t_t [autograd catch all]
```
Huh, I didn't realize "autograd" was a thing when we added printing logic for dispatch table. What does the "autograd" mean here?

ailzhang replied:
Ah it means "autograd kernel is fetched from catch all"... actually it's not necessary I'll just change it to "catch all". :P

ezyang commented on test/test_dispatch.py:
```diff
@@ -337,6 +337,73 @@ def extract_entries(table, dispatch_keys):
+CUDA: default_def_name_t_t [catch all]
+XLA: default_def_name_t_t [catch all]
+AutogradOther: default_def_name_t_t [autograd catch all]
+AutogradCPU: fallthrough registered at ../aten/src/ATen/core/VariableFallbackKernel.cpp:39 [backend fallback]
```
You need to mask this file name / line number out, otherwise it will keep wobbling if someone modifies VariableFallbackKernel.cpp

ezyang commented on test/test_dispatch.py:
```diff
@@ -337,6 +337,73 @@ def extract_entries(table, dispatch_keys):
+            # m.impl("foo", torch::kCPU, [](const Tensor & x) { return x })
+            lambda m: m.impl_t_t("foo", "cpu"),
+            # m.impl("foo", torch::kAutograd, [](const Tensor & x) { return x })
+            lambda m: m.impl_t_t("foo", "autograd"),
```
You should explicitly override the `debug=` kwarg here to give these two functions different names; this way you will be able to tell that they are different in the print below. Right now, they are both named `impl_t_t` and you can't distinguish them in the output.

ezyang reviewed (approved):
Looks good, thanks. Be sure to fix the `debug=` thing and the VariableFallbackKernel.cpp masking problem.

Commit:
Update on "Update precompute logic to skip autograd backend when there's direction"


registration to backend

Differential Revision: [D23589807](https://our.internmc.facebook.com/intern/diff/D23589807)

[ghstack-poisoned]

ailzhang commented:
Ehh fyi committed message in master has wrong title due to commit reordering, but it shouldn't matter since code is correct but just message was for a future PR.
